File: part1/QuickFind/QuickFind.java
import java.util.Arrays;

public class QuickFind {

    private int[] id;

    public QuickFind(int N) {
        id = new int[N];
        for (int i = 0; i < N; i++) {
            id[i] = i;
        }
    }

    public boolean connected(int p, int q) {
        return id[p] == id[q];

    }

    public void union(int p, int q) {
        int pid = id[p];
        int qid = id[q];

        for (int i = 0; i < id.length; i++) {
            if (id[i] == pid) {
                id[i] = qid;
            }
        }
    }

    public static void main(String[] args) {
        QuickFind qf = new QuickFind(10);
        qf.union(4, 3);
        System.out.println(Arrays.toString(qf.id));
        qf.union(3, 5);
        System.out.println(Arrays.toString(qf.id));
        qf.union(5, 6);
        System.out.println(Arrays.toString(qf.id));
        System.out.println(qf.connected(1, 9));
        System.out.println(qf.connected(5, 6));
    }
}

File: part1/QuickFind/QuickFind.js
class QuickFind {
  id = [];

  constructor(N) {
    this.id = new Array(N);
    for (let i = 0; i < N; i++) {
      this.id[i] = i;
    }
  }

  connected(p, q) {
    return this.id[p] === this.id[q];
  }

  union(p, q) {
    const pid = this.id[p];
    const qid = this.id[q];
    for (let i = 0; i < this.id.length; i++) {
      if (this.id[i] === pid) {
        this.id[i] = qid;
      }
    }
  }
}

const qf = new QuickFind(10);
qf.union(4, 3);
qf.union(3, 8);
qf.union(6, 5);
qf.union(9, 4);
qf.union(2, 1);
console.log(qf.connected(8, 9));
console.log(qf.connected(5, 0));
qf.union(5, 0);
qf.union(7, 2);
qf.union(6, 1);
console.log(qf.id);
console.log(qf.connected(8, 9));
